File: algorithms/2_medium/M/00481_magical_string/solution.cpp
class Solution {
public:
    int magicalString(int n) {
        if (n <= 3) {
            return 1;
        }

        const size_t target_len{ static_cast<size_t>(n) };
        string s;
        s.reserve(target_len + 1);
        s += "122";
        size_t index{ 2 };
        char curr_char{ '1' };

        while (s.size() < target_len) {
            s.push_back(curr_char);

            if (s[index] == '2' && s.size() < target_len) {
                s.push_back(curr_char);
            }

            curr_char = curr_char == '1' ? '2' : '1';
            index += 1;
        }

        return std::count(s.cbegin(), s.cend(), '1');
    }
};

File: algorithms/2_medium/M/00481_magical_string/solution.py
class Solution:
    def magicalString(self, target_len: int) -> int:
        if target_len <= 3:
            return 1

        res = [1, 2, 2]
        index = 2
        curr_num = 1

        while len(res) < target_len:
            res.append(curr_num)

            if res[index] == 2 and len(res) < target_len:
                res.append(curr_num)

            curr_num = 1 if curr_num == 2 else 2
            index += 1

        return res.count(1)
